nuxeo-features/nuxeo-platform-oauth/src/test/java/org/nuxeo/ecm/platform/oauth/tests/OAuth2ChallengeFixture.java

```diff
@@ -290,15 +277,51 @@ public void shouldRetrieveAccessAndRefreshToken() throws IOException {
         assertNotSame(refreshed.get("access_token"), token.get("access_token"));
     }
 
-    protected ClientResponse responseFromTokenWith(Map<String, String> queryParams) {
-        WebResource wr = client.resource(BASE_URL).path("oauth2").path(ENDPOINT_TOKEN);
+    protected AuthorizationRequest initValidAuthorizationRequestCall() {
+        Map<String, String> params = new HashMap<>();
+        params.put(CLIENT_ID_PARAM, CLIENT_ID);
+        params.put(RESPONSE_TYPE_PARAM, CODE_RESPONSE_TYPE);
+        params.put(REDIRECT_URI_PARAM, REDIRECT_URI);
+        params.put(STATE_PARAM, STATE);
 
-        MultivaluedMap<String, String> params = new MultivaluedMapImpl();
-        for (Map.Entry<String, String> entry : queryParams.entrySet()) {
-            params.add(entry.getKey(), entry.getValue());
-        }
+        ClientResponse cr = responseFromGetAuthorizationWith(params);
+        assertEquals(200, cr.getStatus());
 
-        return wr.queryParams(params).get(ClientResponse.class);
+        // get back the authorization request from the store for the needed authorization key
+        Set<String> keys = store.keySet();
+        assertEquals(1, keys.size());
+        String key = keys.toArray(new String[0])[0];
+        return AuthorizationRequest.get(key);
+    }
+
+    protected String getAuthorizationCode(String key) {
+        // get an authorization code
+        Map<String, String> params = new HashMap<>();
+        params.put(NuxeoOAuth2Servlet.AUTHORIZATION_KEY, key);
+        params.put(NuxeoOAuth2Servlet.GRANT_ACCESS_PARAM, "true");
+        params.put(STATE_PARAM, STATE);
+        ClientResponse cr = responseFromPostAuthorizationWith(params);
+        assertEquals(302, cr.getStatus());
+        String redirect = cr.getHeaders().get("Location").get(0);
+        String state = extractParameter(redirect, STATE_PARAM);
+        assertEquals(STATE, state);
+        String code = extractParameter(redirect, AUTHORIZATION_CODE_PARAM);
+
+        // ensure we have only one authorization request
+        Set<String> keys = store.keySet();
+        assertTrue(keys.contains(code));
+        assertFalse(keys.contains(key));
+
+        return code;
+    }
+
+    protected String extractParameter(String url, String parameterName) {
+        Pattern pattern = Pattern.compile(parameterName + "=(.*?)(&|$)");
+        Matcher matcher = pattern.matcher(url);
+        if (matcher.find()) {
+            return matcher.group(1);
```

Comment:
For a unit test I guess it's ok but in theory you should unescape the result as well (for url param escaping).